### PR TITLE
Query: Fixes InvalidOperationException on merge to a single partition

### PR DESCRIPTION
**Pull request by neildsh**
## Description

Currently, the cross partition query pipeline components assume that we will always have more than one partition, and throw an InvalidOperationException if they encounter a merge down to a single partition

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs [hunk 1]
```diff
@@ -177,7 +177,7 @@ public async ValueTask<bool> MoveNextAsync(ITrace trace)
                                 this.cancellationToken);
                         }
 
-                        if (childRanges.Count() <= 1)
+                        if (childRanges.Count() < 1)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:180]: nit: childRanges is a list so just use the Count property. Please do this to the rest of places too.

```suggestion
                        if (childRanges.Count < 1)
```

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs [hunk 2]
```diff
@@ -177,7 +177,7 @@ public async ValueTask<bool> MoveNextAsync(ITrace trace)
...
                         {
                             throw new InvalidOperationException("Expected more than 1 child");
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:182]: Not related to this change, but throwing InvalidOperationException is bad. It does not include any of the diagnostics or information necessary to root cause why there was no children.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs [hunk 3]
```diff
@@ -20,24 +20,31 @@ public sealed class CrossPartitionPartitionRangeEnumeratorTests
         [TestMethod]
         public async Task Test429sAsync()
         {
-            Implementation implementation = new Implementation();
+            Implementation implementation = new Implementation(false);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:23]: Did you manually validate that split and merge still work correctly after this change? Can we add this to the CTL runs?

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs (hunk 2 above)
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:182]: Can we add some more context on the error message at least? In this case the number of child ranges was 0, so the current text is not valid anymore. 
Would adding the range help troubleshoot?

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 4]
```diff
@@ -312,7 +312,7 @@ private static class Expressions
                     this.cancellationToken);
             }
 
-            if (childRanges.Count() <= 1)
+            if (childRanges.Count() < 1)
             {
                 throw new InvalidOperationException("Expected more than 1 child");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:317]: Same comment on the error message here.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs (hunk 3 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:23]: Manual validation should at least be ideal. CTL runs don't do merges.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs [hunk 5]
```diff
@@ -49,18 +56,89 @@ public async Task TestEmptyPages()
...
+                    }
+
+                    ++iteration;
+                    if (iteration == 1)
+                    {
+                        this.ShouldMerge = TriState.Ready;
+                    }
+                }
+
+                Assert.AreEqual(numItems, identifiers.Count);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:139]: Are we certain the items are spread across all ranges?

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs (hunk 2 above)
Comment (neildsh) [reply Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs]: throws CosmosException instead

Comment (neildsh) [reply Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs]: fixed error message

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs (hunk 5 above)
Comment (neildsh) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:139]: Yes, undoing the change reliably runs into the exception with this test

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 4 above)
Comment (neildsh) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs]: Fixed the error message

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs (hunk 3 above)
Comment (neildsh) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:23]: This test is the one that exposed the bug: [https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB?path=%2FProduct%2FTests%2FManagement%2FPartitionMergeTestsV3.cs](https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB?path=%2FProduct%2FTests%2FManagement%2FPartitionMergeTestsV3.cs)

It does a merge and will function as additional validation

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs [hunk 6]
```diff
@@ -177,9 +173,17 @@ public async ValueTask<bool> MoveNextAsync(ITrace trace)
...
 
-                        if (childRanges.Count() <= 1)
+                        if (childRanges.Count < 1)
                         {
-                            throw new InvalidOperationException("Expected more than 1 child");
+                            string errorMessage = "SDK invariant violated: Must have at least one EPK range in a cross partition enumerator";
+                            throw new CosmosException(
+                                statusCode: HttpStatusCode.InternalServerError,
+                                message: errorMessage,
+                                stackTrace: string.Empty,
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:182]: Pass null for the stack trace

File: Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs [hunk 7]
```diff
@@ -177,9 +173,17 @@ public async ValueTask<bool> MoveNextAsync(ITrace trace)
                                 this.cancellationToken);
                         }
 
-                        if (childRanges.Count() <= 1)
+                        if (childRanges.Count < 1)
                         {
-                            throw new InvalidOperationException("Expected more than 1 child");
+                            string errorMessage = "SDK invariant violated: Must have at least one EPK range in a cross partition enumerator";
+                            throw new CosmosException(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Pagination/CrossPartitionRangePageAsyncEnumerator.cs:179]: Use the factory:
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/30b3d4e242c0081e2f8a51926bb42511b633daa2/Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosExceptionFactory.cs#L253

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 8]
```diff
@@ -312,9 +308,17 @@ private static class Expressions
                     this.cancellationToken);
             }
 
-            if (childRanges.Count() <= 1)
-            {
-                throw new InvalidOperationException("Expected more than 1 child");
+            if (childRanges.Count < 1)
+            {
+                string errorMessage = "SDK invariant violated: Must have at least one EPK range in a cross partition enumerator";
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:313]: Please use custom error message to distinguish which of the two place it throws incase the user does not capture the stack trace.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/CrossPartitionPartitionRangeEnumeratorTests.cs:23]: That test will not get ran until the next sync. Manual validation is required to validate the fix.

**Review by sboshra — APPROVED** [review]

**Pull request merged**